### ISSUE TITLE
`company-tng`: when the only candidate, <TAB> results in superfluous suffix overlay

### ISSUE THREAD

**Issue by Alexander-Shukaev** [company-mode/company-mode#852]
![output-2018-12-21-02 48 36](https://user-images.githubusercontent.com/911281/50320271-de8a7980-04cb-11e9-9098-e67cbcd95a12.gif)

**Comment by Alexander-Shukaev** [comment]
Maybe it's not about being the last candidate but rather has something to do with quickly hitting `<TAB>`.  Maybe some race condition with `company-idle-delay`?

![output-2018-12-21-02 48 09](https://user-images.githubusercontent.com/911281/50320379-61133900-04cc-11e9-941e-953990cc3f7d.gif)

**Comment by dgutov** [comment]
@nikital Could you take a look?

**Comment by Alexander-Shukaev** [comment]
```
company is a dependency package.

     Status: Installed in ‘company-20181213.2321/’ (unsigned).
    Version: 20181213.2321
    Summary: Modular text completion framework
   Requires: emacs-24.3
Required by: company-ycmd-20180520.1053, company-php-20181110.303, company-flx-20180103.518
   Homepage: http://company-mode.github.io/
   Keywords: abbrev convenience matching 
Other versions: 20181213.2321 (melpa), 0.9.9 (melpa.stable), 0.9.9 (gnu).
```

**Comment by nikital** [comment]
I think it happens because in addition to `company-tng-frontend` you have `company-preview-frontend` in your `company-frontends`. These two  frontends try to "preview" the candidate in different incompatible ways, hence the duplication. Can you try removing the other frontend?

BTW, the recommended configuration to try `company-tng` is `(company-tng-configure-default)`.

**Comment by Alexander-Shukaev** [comment]
You're right, thank you, that was the only problem.  For what's it worth, would it be reasonable to implement, say, a standardized symbol property list for `company` front ends called `conflicts`, where each frontend implemented would be responsible to push symbols of other `company` front ends which could potentially conflict with the developed one?  For example, `(put 'company-tng-frontend 'conflicts '(company-preview-frontend)`, while `company` itself would rigorously check those conflicts upon initialization.

**Comment by dgutov** [comment]
@Alexander-Shukaev Did you change `company-frontends` through Customize?

If so, it might be enough to handle `company-tng-frontend` in `company-frontends-set`.

**Comment by Alexander-Shukaev** [comment]
Nope, all my configurations are done via `setq-default` (as most people probably do anyway).

**Comment by nikital** [comment]
@dgutov Would it be reasonable to specifically add code to company-preview to disable itself if company-tng is active?

On one hand it adds coupling between frontends but on the other hand since in a certain sense preview is logically a "subset" of company-tng (both preview candidates, but tng's preview is more intrusive) then it might give a more intuitive user experience when the user just adds company-tng to the frontend list.

**Comment by dgutov** [comment]
Since `company-tng` is more of an an addon itself, maybe add a check for `company-preview-frontend`'s presence in its code somewhere? And issue a warning about incompatibility when so.

**Comment by Alexander-Shukaev** [comment]
Based on last comment, should this be re-tagged as `enhancement`?

**Comment by dgutov** [comment]
Sure.
